functions/convert/main.py

```diff
@@ -0,0 +1,161 @@
+# -*- coding: utf-8 -*-
+
+# Method for converting MD files to HTML
+
+from __future__ import print_function, unicode_literals
+
+import codecs
+import os
+import re
+import shutil
+import tempfile
+from contextlib import closing
+import markdown
+import boto3
+import datetime
+
+from glob import glob
+from general_tools.file_utils import unzip, add_file_to_zip, make_dir, write_file
+from general_tools.print_utils import print_error, print_ok
+from general_tools.url_utils import download_file
+
+class TransformOBS(object):
+
+    dirRe = re.compile(r'(<div\s.*?class=".*?obs-content.*?">).*?(</div>)', re.UNICODE + re.DOTALL)
+
+    def __init__(self, source_url, outputDir, options):
+        self.source_url = source_url
+        self.outputDir = outputDir
+        self.options = options
+
+        self.downloadDir = tempfile.mkdtemp(prefix='download_')
+        self.filesDir = tempfile.mkdtemp(prefix='files_')
+        self.errors = []
+
+    def close(self):
+        # delete temp files
+        if os.path.isdir(self.downloadDir):
+            shutil.rmtree(self.downloadDir, ignore_errors=True)
+        if os.path.isdir(self.filesDir):
+            shutil.rmtree(self.filesDir, ignore_errors=True)
+
+    def run(self):
+        try:
+            # download the archive
+            fileToDownload = self.source_url
+            filename = self.source_url.rpartition('/')[2]
+            downloadedFile = os.path.join(self.downloadDir, filename)
+            try:
+                print('Downloading {0}...'.format(fileToDownload), end=' ')
+                if not os.path.isfile(downloadedFile):
+                    download_file(fileToDownload, downloadedFile)
+            finally:
+                print('finished.')
+
+            # unzip the archive
+            try:
+                print('Unzipping...'.format(downloadedFile), end=' ')
+                unzip(downloadedFile, self.filesDir)
+            finally:
+                print('finished.')
+
+            # create output directory
+            make_dir(self.outputDir)
+
+            # read the markdown files and output html files
+            try:
+                print('Processing the OBS markdown files')
+
+                files_to_process = sorted(glob(os.path.join(self.filesDir, '*.md')))
+
+                currentDir = os.path.dirname(os.path.realpath(__file__))
+                with codecs.open(os.path.join(currentDir, 'template.html'), 'r', 'utf-8-sig') as html_file:
+                    html_template = html_file.read()
+
+                completeHtml = ''
+                for filename in files_to_process:
+                   # read the markdown file
+                    with codecs.open(filename, 'r', 'utf-8-sig') as md_file:
+                        md = md_file.read()
+                        html = markdown.markdown(md)
+                        completeHtml += html
+                        html = TransformOBS.dirRe.sub(r'\1\n' + html + r'\n\2', html_template)
```

Comment:
@ethantkoenig I prefer to have the template as a separate file. The major benefit is being able to easily validate the html at design time. This takes advantage of some of the major benefits offered by modern IDE's like IntelliJ.